studies/cccl/_segment_algorithms.py

```diff
@@ -0,0 +1,297 @@
+# Copyright (c) 2025, NVIDIA CORPORATION & AFFILIATES. ALL RIGHTS RESERVED.
+#
+#
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+import cupy as cp
+import numpy as np
+import nvtx
+
+from cuda.compute import (
+    PermutationIterator, ZipIterator, TransformIterator, CountingIterator,
+    OpKind, exclusive_scan, select, unary_transform
+)
+
+
+@nvtx.annotate("segment_sizes")
+def segment_sizes(offsets):
+    """
+    Compute the size of each segment from segment offsets.
+
+    Args:
+        offsets: Device array of segment offsets (length = num_segments + 1).
+                 Each segment i contains elements from offsets[i] to offsets[i+1].
+
+    Returns:
+        Device array of segment sizes (length = num_segments).
+    """
+    return offsets[1:] - offsets[:-1]
+
+
+@nvtx.annotate("offsets_to_segment_ids")
+def offsets_to_segment_ids(offsets, stream=None):
+    """
+    Convert segment offsets to segment IDs (indicators).
+
+    Given offsets [0, 2, 5, 8, 10], produces [0, 0, 1, 1, 1, 2, 2, 2, 3, 3]
```

Comment:
Ah, this is what we call `parents` in our kernels :-)

Reply:
That's useful to know when reading Awkward code - thank you!